perf(transformer): faster parsing JSX pragmas from comments

overlookmotel opened the pull request:
Parse JSX pragmas faster by:

1. Finding patterns with byte-by-byte iteration, with fast path for ASCII.
2. Checking for the `@jsx` prefix first before parsing the rest of the pragma.

These routines could be speeded up further by using unchecked slicing (unsafe code). But I've avoided that as the routines at present are quite complex, so hard to verify. The exception is `trim_end`, which is trivial to verify, so I have used unchecked slicing there.

The perf gain here is only ~0.5%, but we should split these routines out into a separate common crate and further optimize them, so that they can be used elsewhere in the codebase where we do string operations, which are often quite expensive.

overlookmotel commented:
* **#6151** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6151?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @overlookmotel and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

Boshen commented on crates/oxc_transformer/src/react/comments.rs:
```diff
@@ -0,0 +1,621 @@
+///
+/// This behavior is aligned with Babel.
+pub(crate) fn update_options_with_comments(options: &mut ReactOptions, ctx: &TransformCtx) {
+    for comment in ctx.trivias.comments() {
```
With the introduction of leading comments, I think iterating through the first set of leading comments should suffice. There are no specifications but I think we can iterate through less comments.

Boshen replied:
Can experiment in another PR.

overlookmotel replied:
Ah OK great!

overlookmotel committed:
perf(transformer): faster parsing JSX pragmas from comments (#6151)

Parse JSX pragmas faster by:

1. Finding patterns with byte-by-byte iteration, with fast path for ASCII.
2. Checking for the `@jsx` prefix first before parsing the rest of the pragma.

These routines could be speeded up further by using unchecked slicing (unsafe code). But I've avoided that as the routines at present are quite complex, so hard to verify. The exception is `trim_end`, which is trivial to verify, so I have used unchecked slicing there.

The perf gain here is only ~0.5%, but we should split these routines out into a separate common crate and further optimize them, so that they can be used elsewhere in the codebase where we do string operations, which are often quite expensive.
